NEW #5711 Add shipment line deleting and editing for draft shipments.

fappels opened the pull request:
# New #5711 Add shipment line deleting and editing for draft shipments.
- Update class for line actions
- Add lot form select
- Add delete button to delete a line
- Add edit button to:
      - edit qty
      - select a new source location
      - select other lot number
      - edit line extrafield

fappels committed:
Merge remote-tracking branch 'refs/remotes/Dolibarr/develop' into develop

fappels committed:
Merge remote-tracking branch 'refs/remotes/Dolibarr/develop' into develop

fappels committed:
Merge remote-tracking branch 'refs/remotes/Dolibarr/develop' into develop

fappels committed:
Add insert, update and delete methods to line class

Cleanup properties
Add insert, line and delete methods

fappels committed:
Add Line triggers, Improve line error handling, ...

Add triggers,
Improve line error handling,
use line->insert in create_line

fappels committed:
Add delete line to expedition card

fappels committed:
Add line id to detail_entrepot object

Add line id to detail_entrepot object.
Add expedition batch id to update batch in line update

fappels committed:
Add entrepot id to expedition batch fetchAll

fappels committed:
Add form select for lot numbers

fappels committed:
Update card for line update

fappels committed:
Fix update extrafield

fappels committed:
Merge remote-tracking branch 'refs/remotes/Dolibarr/develop' into 7.0_new_shipment_update

fappels committed:
Fix delete multi warehouse line

fappels committed:
cleanup

eldy commented on htdocs/expedition/class/expeditionbatch.class.php:
```diff
@@ -207,6 +210,7 @@ static function fetchAll($db,$id_line_expdet)
 				$tmp->fk_origin_stock = $obj->fk_origin_stock;
 				$tmp->fk_expeditiondet = $obj->fk_expeditiondet;
 				$tmp->dluo_qty = $obj->qty;
+				$tmp->entrepot_id = $obj->fk_entrepot;
```
This data come from a link that is ok just after the reception is done, but as soon as stock is empty, the line into product_batch and product_stock will no exists anymore and we won't be able to get the entrepot_id value for old reception. Is this ok for you ?

fappels replied:
I will revert this, thought I need the entrepot_id to easy get the warehouse source of the lot ready to ship, but I seem to have it resolved it in another way. I get it now with a fetch from product_batch. I need the warehouse from product_batch to get it's source location, you should not remove stock before validating. If you remove the lot from stock you also won't be able to validate the shipment. This is also an advantage of the new edit option. When you get the error the lot isn't available anymore, you can replace it by another lot for validating.

eldy commented on htdocs/product/class/html.formproduct.class.php:
```diff
@@ -329,5 +330,126 @@ function load_measuring_units($name='measuring_units', $measuring_style='', $def
+			while ($i < $num)
+			{
+				$obj = $this->db->fetch_object($resql);
+				$this->cache_lot[$obj->rowid]['id'] =$obj->rowid;
```
If fk_product or fk_entrepot are set at first call, we load data restricted to product or warehouse. And the cache array contains only this restricted data.
If we wall again the method loadLotStock with empty fk_product and empty fk_entrepot, then it return the cache. But this cache does not contains all data, so content is not complete for the second call. I suggest to add a key into cache
$this->cache_lot[fk_product.'_'.fk_warehouse]

fappels replied:
I will check this, but will need to consider memory/performance. A company can have a lot off different lot/serial numbers in stock.

fappels replied:
I have added a new parameter to be able to cash all lot stock for all products in lines. So you could also use this new form select for creating lines. I don't see any use case of loading all lot numbers for all products.

fappels committed:
Revert "Add entrepot id to expedition batch fetchAll"

This reverts commit 5342785a4973c1e3d35be59704ec77f399876e1f.

fappels committed:
Fix travis

BebZ commented:
"When you get the error the lot isn't available anymore, you can replace it by another lot for validating."
this is such a major improvement, Francis is the new warehouse's guru :)
"Add shipment line deleting and editing for draft shipments"
> this is exactly the way a shipment must be edited, in a draft it's perfect. 
Just 2 cents in a comment : 
After validation, it means the products left the warehouse (so must be invoiced),
so it should be corrected when still in draft, shipment deletion should be really exceptionnal and recorded in blockchain (regarding fiscal regulations and "validations" of versions regarding laws to respected, in France).

fappels committed:
Only update new lot number

fappels committed:
Add parameter to load all lot from lines of object in cache

If you want use the lot stock selector in mutiple lines you can add
lines array. If you don't give a product id or a lines array, no lot
numbers will be loaded.

fappels committed:
Only show show lot from same src warehouse.

When shipping from multiple warehouse, only allow lot from same src
warehouse. (expeditiondet_batch is on to many from expeditiondet).

fappels committed:
Fix travis

fappels committed:
Merge remote-tracking branch 'refs/remotes/Dolibarr/develop' into 7.0_new_shipment_update

fappels committed:
Fix create line extra fields
